### PR TITLE
WIP: Use _Brain in plot_vector_source_estimates

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
This PR is a follow-up of #6731 suggested also in [this comment](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-536042199). The goal is to make `stc_vec.plot()` use the `_Brain` class which relies on the abstract 3d backend layer `_Renderer`.

**ToDo**

- [x] Fix scaling of the glyphs by using scalar values (merged [upstream](https://github.com/pyvista/pyvista/pull/451))
- [x] Enable Depth Peeling (merged [upstream](https://github.com/pyvista/pyvista/pull/450))
- [x] Increase arrow thickness
- [x] Fix colorbar range
- [x] Update the 3D tests with the `renderer` fixture (c.f. https://github.com/mne-tools/mne-python/pull/7028#issue-337436335)
- [x] Experiment 3D arrows glyphs (feedback from #7030)
- [x] Colormap has a 'grey tone'
- [x] Refactor smoothing matrix generation
- [x] The left and right hemi data have different scales
- [x] Rendering dark filter bug on CircleCI
- [x] remove glyph options (should be future PR)
- [x] `time_viewer=True` only playback only updates mesh, not vectors
- [x] hitting spacebar at the end of playback aborts Python (should restart from beginning), fix was in the reverted PR
- [x] fix glyph flashing during playback (if still an issue)
- [x] get CIs green
- [x] add support for `show_traces=True` if it's not too difficult
- [x] fix smoothing steps bug

It's an item of #7162

### CODE REVIEW

**Comment by GuillaumeFavelier** [comment]
This is my local result using the [gist](https://gist.github.com/rythorpe/c51a5b66b492b67437843283c67050af) suggested in [this comment](https://github.com/mne-tools/mne-python/issues/7020#issue-517400092). This is still work in progress.

2019-11-14_1920x1080

**Comment by GuillaumeFavelier** [comment]
Better to wait for the merge of #7084, I expect some conflicts to solve.

**Comment by GuillaumeFavelier** [comment]
Meanwhile, these are my results: 

Mayavi | PyVista
---------|----------
2019-11-20_1920x1080 | 2019-11-20_1920x1080_000

Note: this "grey tone" is still here. I suspect there is some scaling of the colormap to do.

**Comment by larsoner** [comment]
Apparently no conflicts after merging #7084, but feel free to rebase anyway @GuillaumeFavelier

**Comment by GuillaumeFavelier** [comment]
I think we don't see the conflicts because Azure is failing. I see them locally though

**Comment by larsoner** [comment]
@GuillaumeFavelier any chance to get to this in the next two weeks to get this into 0.20? That way we can have a complete replacement for 3D viz (outside of `mne coreg`), right?

**Comment by GuillaumeFavelier** [comment]
I'll increase the priority of this PR but after so long the same blockers remain:

* The glyph size is not consistent with Mayavi
* Some kind of grey filter is applied on the colorbar
* The renders on sphinx-gallery are far from satisfying

**Comment by larsoner** [comment]
Okay I'll move the milestone to 0.21 then

**Comment by GuillaumeFavelier** [comment]
Eventually I think this could be merged but the overview table in `mne/viz/backends/renderer.py` should be updated with only partial support for `plot_vector_source_estimates` with an additional note saying what's missing or not working.

**Comment by GuillaumeFavelier** [comment]
I notice a very big lag during playback when `time_viewer=True` and I think the memory consumption is huge especially when `hemi='both'`. Did I do something wrong with the interpolation?

**Comment by larsoner** [comment]
Is it spatially upsampling the entire n_vertices x n_times array (bad), or upsampling a single time point at a time (good) ?

**Comment by GuillaumeFavelier** [comment]
I suppose it's bad so I'll try "upsampling a single time point at a time"

**Comment by larsoner** [comment]
> I suppose it's bad so I'll try "upsampling a single time point at a time"

It looked like the only place the upsampling of the full was being done was to get the max -- and this you can compute on the data itself. Pushed a commit to do this. Feel free to try it and see if things are still laggy. Hopefully they aren't!

**Comment by larsoner** [comment]
Also pushed a commit to touch probably relevant examples

**Comment by larsoner** [comment]
I'm going to re-mark this for 0.20 and we can scrap it if need be

**Comment by larsoner** [comment]
Looks like it's getting close @GuillaumeFavelier! Updated examples:

- tutorials/source-modeling/plot_dipole_orientations.py [PR](https://18783-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/source-modeling/plot_dipole_orientations.html#loose-dipole-orientations) / [master](https://mne.tools/dev/auto_tutorials/source-modeling/plot_dipole_orientations.html#loose-dipole-orientations)
- tutorials/source-modeling/plot_visualize_stc.py [PR](https://18783-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/source-modeling/plot_visualize_stc.html#vector-source-estimates) / [master](https://mne.tools/dev/auto_tutorials/source-modeling/plot_visualize_stc.html#vector-source-estimates)
- examples/inverse/plot_vector_mne_solution.py [PR](https://18783-1301584-gh.circle-artifacts.com/0/dev/auto_examples/inverse/plot_vector_mne_solution.html#sphx-glr-auto-examples-inverse-plot-vector-mne-solution-py) / [master](https://mne.tools/dev/auto_examples/inverse/plot_vector_mne_solution.html#sphx-glr-auto-examples-inverse-plot-vector-mne-solution-py)

By eye all three of these are getting close but suffer from the same problems:

- some rendering problems, maybe blending modes or something, probably worth checking what PySurfer does?
- colorbar is wrong

**Comment by GuillaumeFavelier** [comment]
> some rendering problems, maybe blending modes or something, probably worth checking what PySurfer does?

I investigated for some time now and I can't seem to reproduce *exactly* what PySurfer does. I noticed the rendering issues too but I have no solution for that.

> colorbar is wrong

I'll see what I can do

**Comment by GuillaumeFavelier** [comment]
Strangely enough, the grey tone on the colorbar is fixed when `time_viewer=True`...

**Comment by GuillaumeFavelier** [comment]
The grey tone and the range of the colorbar are fixed now.

**Comment by GuillaumeFavelier** [comment]
Regression detected in: 

https://18790-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/source-modeling/plot_dipole_orientations.html#sphx-glr-auto-tutorials-source-modeling-plot-dipole-orientations-py

I have to fix this

**Comment by GuillaumeFavelier** [comment]
The order of certain drawing operations are important to achieve the correct render, I commented to make it clear.

**Comment by larsoner** [comment]
Style is unhappy

Does it render properly now? If so, what else is left to do?

File: mne/viz/_brain/_brain.py [hunk 1]
```diff
@@ -453,6 +450,36 @@ def time_label(x):
...
+            if isinstance(mesh_data, tuple):
+                actor, mesh = mesh_data
+                # add metadata to the mesh for picking
+                mesh._hemi = hemi
+            else:
+                actor, mesh = mesh_data, None
+            self._data[hemi]['actor'].append(actor)
+            self._data[hemi]['mesh'].append(mesh)
+
+        # 2) udpate time and smoothing properties
```
Comment (larsoner) [inline mne/viz/_brain/_brain.py]: udpate -> update

**Comment by GuillaumeFavelier** [comment]
The colorbar is fixed, this:

> some rendering problems, maybe blending modes or something

is not fixed. I updated the note about it in `mne/viz/backends/renderer`

**Comment by GuillaumeFavelier** [comment]
Waiting on the CIs to confirm

**Comment by larsoner** [comment]
> is not fixed. I updated the note about it in mne/viz/backends/renderer

Can you update the top comment? It seems like that's out of date now. It would be good to know what still needs to be done. I might have time to look into the rendering bugs today

**Comment by GuillaumeFavelier** [comment]
It's updated now @larsoner

**Comment by larsoner** [comment]
After having read through the code that PySurfer uses, I'm convinced it's wrong because it scales the left and right hemisphere data potentially differently, unless it's compensating for something that Mayavi does under the hood. This will require more investigation, but if it's the case we should probably fix the behavior somehow :(

**Comment by larsoner** [comment]
In the meantime I pushed a commit that unifies how/where smoothing matrices are calculated by refactoring slightly and reusing code from `mne/morph.py` rather than `mne/viz/_brain/utils.py`

File: mne/viz/_brain/_brain.py [hunk 2]
```diff
@@ -486,15 +487,6 @@ def time_label(x):
                 vertices = slice(None) if vertices is None else vertices
                 x, y, z = np.array(self.geo[hemi].coords)[vertices].T
 
-                if scale_factor is None:
-                    width = np.ptp(self.geo[hemi].coords[:, 1])
-                    final_scale_factor = width * 0.1
```
Comment (GuillaumeFavelier) [inline mne/viz/_brain/_brain.py]: We could iterate through the hemis here to find a suitable scale factor

**Comment by GuillaumeFavelier** [comment]
With the latest version of `quiver3d()`, the following script gives the same result for both `mayavi` and `pyvista`:

```py
import mne

for backend in ('mayavi', 'pyvista'):
    mne.viz.set_3d_backend(backend)
    from mne.viz.backends.renderer import _get_renderer
    # scale by vector values
    renderer = _get_renderer()
    renderer.quiver3d([0, 0, 0], [0, 0, 0], [0, 0, 0],
                      [0, 0, 5], [10, 0, 0], [0, 25, 0],
                      color="red", scale=2., scale_mode='vector',
                      mode="2darrow")
    renderer.show()

    # scale by scalar values
    renderer = _get_renderer()
    renderer.quiver3d([0, 0, 0], [0, 0, 0], [0, 0, 0],
                      [0, 0, 1], [1, 0, 0], [0, 1, 0],
                      scalars=[10, 25, 5],
                      color="red", scale=2., scale_mode='scalar',
                      mode="2darrow")
    renderer.show()
```

**Comment by agramfort** [comment]
ready to test?

>

**Comment by GuillaumeFavelier** [comment]
I'm still not satisfied by this version @agramfort. I tried to work directly on `quiver3d` to obtain something backend-agnostic but either in the parameters sent by `PySurfer` or the post-processing on the glyphs, there is something that remains mysterious to me that I can't quite reproduce with `PyVista`.

**Comment by GuillaumeFavelier** [comment]
I believe 8073aef should reproduce https://github.com/nipy/PySurfer/blob/master/surfer/viz.py#L2049-L2061 and fix the color issues on the glyphs.

**Comment by larsoner** [comment]
@GuillaumeFavelier let me know when you're tired of trying and I'll take a look

**Comment by GuillaumeFavelier** [comment]
@larsoner I pushed a commit to fix the warnings in the CIs, feel free to take a look

**Comment by GuillaumeFavelier** [comment]
@larsoner I pushed a commit to enable depth peeling on pyvista. It's especially useful for the users of `vtk<8.2` because after this, it's enabled by default

File: mne/viz/_3d.py [hunk 3]
```diff
@@ -2365,6 +2365,10 @@ def plot_vector_source_estimates(stc, subject=None, hemi='lh', colormap='hot',
     time_unit : 's' | 'ms'
         Whether time is represented in seconds ("s", default) or
         milliseconds ("ms").
+    glyph : '2darrow' | '3darrow'
+        The type of glyphs to use. If '2darrow', the glyphs belong to a plan
+        and if '3darrow', the glyphs are composite 3d objects (cone +
+        cylinder).
```
Comment (larsoner) [inline mne/viz/_3d.py]: @GuillaumeFavelier can we just get rid of this option for now? I'm not sure it's going to work given all the scaling issues

Comment (GuillaumeFavelier) [reply mne/viz/_3d.py]: It should not be difficult to add later. I'll remove it for now because it's not top priority

**Comment by GuillaumeFavelier** [comment]
The flickering of the glyphs happens because the glyphs actor is removed and re-added in the scene. So if I update the geometry of the glyphs already present in the scene without adding a new actor, it should be fixed.

**Comment by larsoner** [comment]
Are you working on that now, or is it implemented already?

**Comment by GuillaumeFavelier** [comment]
I'm pushing right now

**Comment by GuillaumeFavelier** [comment]
This is what I obtain:

output

**Comment by larsoner** [comment]
Awesome!

Want to merge with master and try show_traces='auto'?

**Comment by GuillaumeFavelier** [comment]
I think this PR fails on azure somehow

**Comment by larsoner** [comment]
Looks like the failure occurs in `test_report.py`, or at least that's the test that should be next in the order. Could something here have broken it?

**Comment by agramfort** [comment]
this works great now for me !

can we fix conflict and merge when CIs are green?

**Comment by GuillaumeFavelier** [comment]
I'll investigate the failures

**Comment by GuillaumeFavelier** [comment]
Waiting on Travis here then I'll work on Azure

File: mne/source_estimate.py [hunk 4]
```diff
@@ -2246,7 +2246,7 @@ def __init__(self, data, vertices=None, tmin=None, tstep=None,
     def plot_surface(self, src, subject=None, surface='inflated', hemi='lh',
                      colormap='auto', time_label='time=%02.f ms',
                      smoothing_steps=10,
-                     transparent=None, alpha=1.0, time_viewer=False,
+                     transparent=None, alpha=1.0, time_viewer='auto',
```
Comment (larsoner) [inline mne/source_estimate.py]: These will also need to be changed in `_3d.py`

Comment (GuillaumeFavelier) [reply mne/source_estimate.py]: This was done in #7380 no?

**Comment by GuillaumeFavelier** [comment]
The failures don't seem related to depth peeling, I'll have to split this PR to have a better perspective

File: mne/source_estimate.py (hunk 4 above)
Comment (larsoner) [reply mne/source_estimate.py]: Ahh, yes I guess it was

**Comment by larsoner** [comment]
Killing CIs for now while we wait for #7513

**Comment by GuillaumeFavelier** [comment]
Okay

**Comment by larsoner** [comment]
... actually I'll just bring the optimistic fix here

**Comment by GuillaumeFavelier** [comment]
Checking Azure here then

**Comment by GuillaumeFavelier** [comment]
It seems it's indeed related to AA. And you think it would not work on Azure if we remove:

```py
    if os.getenv('AZURE_CI_WINDOWS', 'false').lower() == 'true':
        return
```

Now that we check if the figure is active?

File: mne/viz/_3d.py [hunk 5]
```diff
@@ -2391,9 +2400,18 @@ def plot_vector_source_estimates(stc, subject=None, hemi='lh', colormap='hot',
...
+                  (True, False, 'auto', 'separate'))
     using_mayavi = get_3d_backend() == "mayavi"
     if time_viewer == 'auto':
         time_viewer = not using_mayavi
+    if show_traces == 'auto':
+        show_traces = (
+            not using_mayavi and
+            time_viewer and
+            # XXX temporary hidden workaround for memory problems on CircleCI
+            os.getenv('_MNE_BRAIN_TRACES_AUTO', 'true').lower() != 'false')
```
Comment (larsoner) [inline mne/viz/_3d.py]: This looks like code dup that should be refactored with `plot_source_estimates`

**Comment by larsoner** [comment]
> It seems it's indeed related to AA. And you think it would not work on Azure if we remove:

The old check was just to see if `ren_win` existed. All the PyVista code does is set some `ren_win` things and call `modified`. So it should have errored out if `ren_win` did not exist.

So I think `is_active` might not be checking that things are sufficiently activated (e.g., OpenGL context actually created) to set the necessary properties.

**Comment by larsoner** [comment]
... oh I guess it does not actually check `ren_win` (the depth peeling code does). So maybe it would be sufficient. I guess you could try it, but it might be better to put effort into getting this PR merged with green CIs, then undo some of the workarounds if possible in later PRs.

**Comment by GuillaumeFavelier** [comment]
> getting this PR merged with green CIs, then undo some of the workarounds if possible in later PRs.

+100, this PR is up for too long

**Comment by GuillaumeFavelier** [comment]
> This looks like code dup that should be refactored with plot_source_estimates

Sorry for being paranoid but I'll push my latest changes once the CIs come back green.

**Comment by larsoner** [comment]
Awesome that `time_viewer='auto'` is now working!

@GuillaumeFavelier another fix that I think was in the reverted PR that I just noticed was that `smoothing_steps` now gets overridden by the slider value instead of the slider being set to whatever value was passed in `smoothing_steps`. You can replicate by passing for example `smoothing_steps=1` which very clearly have gaps, it shows up then gets "snapped" up to 7 steps.

Also this suggests that the steps are being computed twice during init, which will slow things down.

If you fix this, please also check that `smoothing_steps='nearest'` works properly, because we have to be careful about converting that to `0` when setting the slider value.

Then I think we're ready for merge, right?

**Comment by GuillaumeFavelier** [comment]
I fixed the bugs with `smoothing_steps` and updated the tests. I also removed the duplicated calls to `set_data_smoothing` and `set_time_point` in `_TimeViewer`.

Looking at the todo list and assuming the CIs come back green, I think we're ready now.

**Comment by agramfort** [comment]
still works great for me !

>

**Comment by GuillaumeFavelier** [comment]
> still works great for me !

Very reassuring @agramfort 😅

**Comment by larsoner** [comment]
Works great! Thanks for the awesome work @GuillaumeFavelier

**Pull request merged**